fixtures/webstudio-remix-vercel/app/routes/[resources]._index.tsx

```diff
@@ -0,0 +1,308 @@
+          format: "raw",
+        })}`,
+      });
+    }
+  }
+
+  for (const customMeta of page.meta.custom ?? []) {
+    if (customMeta.property.trim().length === 0) {
+      continue;
+    }
+    metas.push(customMeta);
+  }
+
+  return metas;
+};
+
+export const links: LinksFunction = () => {
+  const result: LinkDescriptor[] = [];
+
+  result.push({
+    rel: "stylesheet",
+    href: css,
+  });
+
+  const { site } = pageData;
+
+  if (site?.faviconAssetId) {
+    const imageAsset = imageAssets.find(
+      (asset) => asset.id === site.faviconAssetId
+    );
+
+    if (imageAsset) {
+      result.push({
+        rel: "icon",
+        href: imageLoader({
+          src: imageAsset.name,
+          width: 128,
+          quality: 100,
+          format: "auto",
+        }),
+        type: undefined,
+      });
+    }
+  } else {
+    result.push({
+      rel: "icon",
+      href: "/favicon.ico",
+      type: "image/x-icon",
+    });
+
+    result.push({
+      rel: "shortcut icon",
+      href: "/favicon.ico",
+      type: "image/x-icon",
+    });
+  }
+
+  for (const asset of fontAssets) {
+    if (asset.type === "font") {
+      result.push({
+        rel: "preload",
+        href: assetBaseUrl + asset.name,
+        as: "font",
+        crossOrigin: "anonymous",
+        // @todo add mimeType
+        // type: asset.mimeType,
+      });
+    }
+  }
+
+  return result;
+};
+
+const getRequestHost = (request: Request): string =>
+  request.headers.get("x-forwarded-host") || request.headers.get("host") || "";
+
+const getMethod = (value: string | undefined) => {
+  if (value === undefined) {
+    return "post";
+  }
+
+  switch (value.toLowerCase()) {
+    case "get":
+      return "get";
+    default:
+      return "post";
+  }
+};
+
+export const action = async ({ request, context }: ActionArgs) => {
+  const formData = await request.formData();
+
+  const formId = getFormId(formData);
+  if (formId === undefined) {
+    // We're throwing rather than returning { success: false }
+    // because this isn't supposed to happen normally: bug or malicious user
+    throw json("Form not found", { status: 404 });
+  }
+
+  const formProperties = formsProperties.get(formId);
+
+  // form properties are not defined when defaults are used
+  const { action, method } = formProperties ?? {};
+
+  const email = user?.email;
+
+  if (email == null) {
+    return { success: false };
+  }
+
+  // wrapped in try/catch just in cases new URL() throws
+  // (should not happen)
+  let pageUrl: URL;
+  try {
+    pageUrl = new URL(request.url);
+    pageUrl.host = getRequestHost(request);
+  } catch {
+    return { success: false };
+  }
+
+  if (action !== undefined) {
+    try {
+      // Test that action is full URL
+      new URL(action);
+    } catch {
+      return json(
+        {
+          success: false,
+          error: "Invalid action URL, must be valid http/https protocol",
```

Comment:
this reads like the error is because of the protocol itself, but is about overall invalid URL, I would change the wording to just "Invalid URL" if we can't tell exactly what's wrong